addons/crm/models/crm_lead.py

```diff
@@ -382,31 +388,56 @@ def create(self, vals_list):
         return leads
 
     def write(self, vals):
-        # stage change:
-        if 'stage_id' in vals:
-            stage_id = self.env['crm.stage'].browse(vals['stage_id'])
-            if stage_id.is_won:
-                vals.update({'probability': 100})
-
-        # stage change with new stage: update probability and date_closed
+        #   stage change with new stage: update probability and date_closed
         if vals.get('probability', 0) >= 100 or not vals.get('active', True):
             vals['date_closed'] = fields.Datetime.now()
         elif 'probability' in vals:
             vals['date_closed'] = False
 
+        if any(field in ['active', 'stage_id'] for field in vals):
+            vals = self._handle_won_lost(vals)
+
         write_result = super(Lead, self).write(vals)
         # Compute new automated_probability (and, eventually, probability) for each lead separately
         if self._should_update_probability(vals):
             self._update_probability()
 
         return write_result
 
+    def _handle_won_lost(self, vals):
+        # Global Won / Lost impacts
+        new_stage = self.env['crm.stage'].browse(vals.get('stage_id'))
+        if new_stage.is_won:
+            vals['probability'] = 100
+
+        # Lead by Lead impact
+        Lead = self.env['crm.lead']
+        leads_reach_won = Lead
+        leads_reach_lost = Lead
+        leads_leave_lost = Lead
+        for lead in self:
+            if new_stage.is_won:
+                if lead.probability == 0:
+                    leads_leave_lost |= lead
+                leads_reach_won |= lead
+            if 'active' in vals:
+                if not vals['active'] and lead.active:  # archive lead
+                    leads_reach_lost |= lead
+                elif vals['active'] and not lead.active:  # restore lead
+                    leads_leave_lost |= lead
+
+        leads_reach_won._pls_reach_state('won')
+        leads_reach_lost._pls_reach_state('lost')
+        leads_leave_lost._pls_leave_state('lost')
+
+        return vals
+
     def _update_probability(self):
         lead_probabilities = self.sudo()._pls_get_naive_bayes_probabilities()
         for lead in self:
             lead_proba = lead_probabilities.get(lead.id, 0)
             proba_vals = {'automated_probability': lead_proba}
-            if lead.is_automated_probability:
+            if lead._origin.id and tools.float_compare(lead.probability, lead.automated_probability, 2) == 0:
```

Comment:
Here you use `_origin.id` and above `lead_probabilities.get(lead.id, 0)` ? Doesn't seem right.

Reply:
that's not the same thing. lead.id is the key to get the corresponding lead_probability. 
Testing origin.id is the check if the record is not new (can ben false in create).
In fact we are checking manually the is_automated_probability. (see fix commit)